### ISSUE TITLE
Add support for Oracle Cloud Infrastructure (OCI) DNS Service

### ISSUE THREAD

**Issue by Djelibeybi** [acmesh-official/acme.sh#3540]
> **NOTE**: support for the Oracle Cloud Infrastructure (OCI) DNS Service has been submitted via pull request #3539. The content of this first post will be published as usage instructions on the wiki and the issue itself will be used for reporting bugs in the plugin after the pull request is merged.

The Oracle Cloud Infrastructure Domain Name System (DNS) service lets you [create and manage public and private DNS zones][DNS] in an Oracle Cloud tenancy. The DNS Service provides each tenancy with a limit of 1000 zones and 25,000 records per zone and is [charged per 1M queries][COST].

## Using Oracle Cloud Infrastructure (OCI) DNS Service with `acme.sh`

Before you can use OCI DNS with `acme.sh`, you'll need  the following:

* A private [API signing key][APIKEY]
* Your [tenancy and service account OCIDs][OCIDS]

The [OCI Developer Guide][DG] covers the procss to generate and retrieve these details.

We recommend [installing the OCI CLI][CLI] as it has an [interactive setup process][CSTP] that walks you through creating an API signing keypair and retrieving the required OCIDs. Once the CLI is installed and configured, you can find the required details in `~/.oci/config` and the private key can be found in `~/.oci/oci_api_key.pem`.

### Configuring the `dns_oci` plugin

The following four environment variables are mandatory:

* `OCI_USER`: OCID of the user calling the API. Example: `ocid1.user.oc1..<unique_ID>`
* `OCI_FINGERPRINT`: Fingerprint for the public key that was added to this user
* `OCI_PRIVATE_KEY`: Path to the file containing the API signing key in PEM format.
* `OCI_TENANCY`: OCID of your tenancy. Example: `ocid1.tenancy.oc1..<unique_ID>`

The following two environment variables are optional:

* `OCI_KEY_PASSPHRASE`: Passphrase used for the key, if it is encrypted.
* `OCI_REGION`: An Oracle Cloud Infrastructure region. See [Regions and Availability Domains][REGS] for a complete list. Defaults to `us-ashburn-1` if unspecified.

#### Other methods of providing an API signing private key

There are two scenarios that require a different method of providing the API signing private key:

* running `acme.sh` from location where storing a private key on-disk would be considered a significant **security risk**; or
* as part of an automation, orchestration or CI/CD implementation which limit variables to **URI-safe strings** that do not contain line breaks or special characters.

The plugin provides an option for each of these scenarios:

##### Example: GitHub Actions (security risk)

To run `acme.sh` as part of a GitHub Actions workflow, store the **required** environment variables as [GitHub encrypted secrets][GHS]. Store the _content_ of the API signing private key as the value of the `OCI_PRIVATE_KEY` secret. This can be done via the Settings section of your GitHub repository or with the [GitHub CLI][GCLI] [`gh secret set`] command:

```shell
$ gh secret set OCI_USER -b"$OCI_USER"
✓ Set secret OCI_USER for my_user/git_repo

$ gh secret set OCI_TENANCY -b"$OCI_TENANCY"
✓ Set secret OCI_TENANCY for my_user/git_repo

$ gh secret set OCI_FINGERPRINT -b "$OCI_FINGERPRINT"
✓ Set secret OCI_FINGERPRINT for MyAccount/GitRepo

$ gh secret set OCI_PRIVATE_KEY < "$HOME/.oci/oci_api_key.pem"
✓ Set secret OCI_PRIVATE_KEY for MyAccount/GitRepo
```

##### Example: `docker run --env-file` (URI-safe string)

The `--env-file` parameter of the `docker run` command [cannot contain multiline environment variables][MOBY12997]. In this scenario, use the `openssl` tool to encode the API signing private key into base64 and provide that as the value of `OCI_PRIVATE_KEY`:

```shell
$ export OCI_PRIVATE_KEY="$(openssl enc -a -A < /path/to/oci_api_key.pem)"
$ echo "$OCI_PRIVATE_KEY"
LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFb2dJQkFBS0NBUUVBdElRNlJLaWxsMmpRakJtTWlMM005b0xtbUhKa2Z4ajg3VklLeDFKRStNdEZPOU91Ckc0RHZ2dmVVTHB3KzBVRXFrbGczb2hnSXp2TXZ5cXhWZS9yYW01aDBuSlNnU0RnY2RBYVNoSDZ6c3BCeUpoQWIKem1IY3VlMHp6TVJzc0ppeVVyMmpRWEtFL3kraFVTa1o5ZjhTR0MvYXR6NnVQOWM3aEhoczd0N2Y3ajlCNkF3cQpkQVFGMFN5NXE5alJMRGFFQ0pxYjJIRm13b2YrN1JDRlB4dEJ4ajI0bnVHTXRJVHZMeEZ4NElWMEJJU2NQRzA3CkJrRndBUlNuZ0NSZ0xneWxhRnBZYVZFVXZqcFJmUFgxdFBkTEV0ampPVjhEUGtET21WSm84OWlsUloxWE9lS0QKRWxIczNlMEsxMTRtdU13NVozaGdWVTdudFRPQ3owQkY2bkx2VFFJREFRQUJBb0lCQUZ4SFUvb1JORVZJUVlPOAo4OGhiTGxHOTdOOTFhc1pHSWRXM29PYnZ1VzV0aXdEZ1h4cVVPUHFvV1VuU2VaSkMrOTQvMW9DNFJKaDJYTElBCmt0cUVEQWlLVU9KU1gvVkxLZEJ1STZ3emtPenhSa2tLNXd3RG52QTB2OHduaFdCcFRvOVpmZ0ZRNTRnRDh6VW8KQzZ0akx2N0tSOFp3OW1hUVFrNmZ1U0FMTXZINDdCT2ZRN3NxRXk5UTBNdDZKSVIvdDQwV3BWRnJLY0ZPL3lYaQpNZFdXN0h0eCtjeEk1K2I1RzQzT0oxTzYxczZjVlZRVkd4Szg5YXBtVEZxaS9jcGNYSGMwYTZOa0wwZWZoQ0FLCmJLTngrRHB1cXEzQW53RGtYV1ZocVRacVp3RmVNZmFnWW5xSEt3NExwOTZCeUVJdkVoeU5Td0tLNDJwcndRYkwKbEZuUGxNRUNnWUVBNWM0eHlIMC96bVVUcFBJdk0wL1N1UVZXcEw0MVUyMFFpLzNNMXliSU9FYzd2eHVBWGdacQpuVENlQktiWW4zR0Z2d2ZnZklPeFA5dUpHMEx1VUMrcXdhSjUrdkl6bTllNXlJVjFMVEpDVVZUUVJRYWtzVU5VCjJMZmovOVpYTWlZM0FXWXhramJzTE1wWGw0ejB1eXhlYkFncGRjM1B2eURtbUFrZlJNVDZtVVVDZ1lFQXlSZkQKclhVQXQ4azdCZDVGempuRDMvR2d0bkRtSmJzSGk3ZGxYWDd4RXhzcG04Q0ZLKzRkUERsMW9rMktrK3JjdVJxNwo2YmNldmVJT254c21yMUlKM3RyeHgra1lPYVdiRlR3S3FJWmlxbVJCcFBjVnZFSnpEcHdocitPQ3lLR0Z3dlJGCmF6Zm4raVF3ektIRksrR1d6YXBobkRUMHJyeGd3aHBCWVRQZDZta0NnWUFrdSswaFMwUWp6ZXh4OVhTdUV4aHcKRXZaaXB6VUcvT3drTkpIRjRxRW12ZHhERkRaVTBWbkJmc0FEcGxnQ1JMZEliZzNvV1hTLzdHdmlPN1MwaEtvYgpBY25WMzFsNGsrZFNzbUtrc3QvKytEdkNqT2xPbktDM2M4MnlXNlRjazFjL2xnSmVpUUQvdCtpQWhUdXBuOUlJCkd0WGx5dHhtaE8vWFpiM2wxRWJxc1FLQmdHQlFSb05rbjdSbTZDTVdZMzZ6Z2hKWGEyOFdRZFRlY0czRnFBSVEKMDZ3Zi9WekE1OGV6bzBMK0RrNXZMeWxMZXU0WlZObEt3S21YaFFORmdOREZ6VDUweFFySHBMZXZqc2lyMmpyLwpLbkNMMWprSnN0cFpSVHhCSVkxWnZZRVVnN2x0UDNacjcycUtQT2sxcmZjbmRCQUxRL2NLRWd0QzVHRHpMRTgvCkJWSEpBb0dBVFhHS2JtTklId3dQdU9TSjFKZ21XSDhvSC9OQUNrOUZwdHUvZ0FTYSt0S1RpSkZ6c0lBVnUxc0QKQkg5dEJ4UUJicTY5NTFHbE96UzFpWjRLRFdMQ2FPZmpGZENRcEtXK2tmRGpwQlBhMlhXZGlXdXFQOG9sdCt4WgpVTW1uWUZHVGh2Q1ppSXA0ejhtN25ZSWNKS0ZGK2xFd2p6ODJqeklDdXYvalVKaXZrZlU9Ci0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg==
```

### Issuing a certificate

To issue a certificate, ensure all the mandatory environment variables have been configured, then run:

```shell
acme.sh --issue --dns dns_oci -d example.com -d www.example.com
```

To issue a wildcard certificate, use::

```shell
acme.sh --issue --dns dns_oci -d example.com -d *.example.com
```

## Security recommendations

The **[Oracle Cloud Infrastructure Security Guide][OSG]** details our recommended
**[best practices for securing user authentication][BP]** which include:

* creating **a dedicated service user account** specifically for GitHub Actions;
* assigning that service account a **unique** and **complex** password;
* **rotating the API signing key pair** used by the service account every 90 days; and
* using **[GitHub encrypted secrets][GHS]** to store credentials.

[DNS]: https://docs.oracle.com/en-us/iaas/Content/DNS/Tasks/managingdnszones.htm#Managing_DNS_Service_Zones
[COST]: https://www.oracle.com/cloud/price-list.html#dns
[DG]: https://docs.oracle.com/en-us/iaas/Content/API/Concepts/devtoolslanding.htm
[APIKEY]: https://docs.oracle.com/en-us/iaas/Content/API/Concepts/apisigningkey.htm
[OCIDS]: https://docs.oracle.com/en-us/iaas/Content/API/Concepts/apisigningkey.htm#five
[CLI]: https://docs.oracle.com/en-us/iaas/Content/API/Concepts/cliconcepts.htm
[CSTP]:https://docs.oracle.com/en-us/iaas/Content/API/SDKDocs/cliinstall.htm#configfile
[OSG]: https://docs.oracle.com/en-us/iaas/Content/Security/Concepts/security_guide.htm
[BP]:  https://docs.oracle.com/en-us/iaas/Content/Security/Reference/iam_security.htm
[GHS]: https://docs.github.com/en/actions/reference/encrypted-secrets
[REGS]: https://docs.oracle.com/en-us/iaas/Content/General/Concepts/regions.htm#top
[MOBY12997]: https://github.com/moby/moby/issues/12997
[GCLI]: https://cli.github.com/

**Comment by Neilpang** [comment]
thanks, I will check it soon.

**Comment by Neilpang** [comment]
I don't think `OCI_FINGERPRINT ` is necessary.
Can we calcualte it from the `OCI_PRIVATE_KEY ` ?

**Comment by Neilpang** [comment]
please also move `OCI_REGION` in to the first section,   It's so common for a user to specify a region.

**Comment by Djelibeybi** [comment]
No, the `OCI_FINGERPRINT` is for the public key that we don't have. And `OCI_REGION` is optional, which is why it's in the second section.

**Comment by Neilpang** [comment]
I knew `OCI_REGION ` is optional,  but there will be many users reporting bug if it's not working for him,  just because they are not in the defalt region.

**Comment by Djelibeybi** [comment]
You don't need to use the default region. Any DNS endpoint will work, because the DNS zones are no longer region specific.

**Comment by Neilpang** [comment]
>> You don't need to use the default region. Any DNS endpoint will work, because the DNS zones are no longer region specific.

I see.

**Comment by Neilpang** [comment]
>> No, the OCI_FINGERPRINT is for the public key that we don't have

Is the public from the `OCI_PRIVATE_KEY ` ?

**Comment by Djelibeybi** [comment]
Yes, we could generate the public key and then the fingerprint, but the fingerprint is something users are used to configuring (it's required for the OCI CLI configuration). I wanted the configuration to feel the same as the CLI, even though we don't use it.

**Comment by Neilpang** [comment]
>>  but the fingerprint is something users are used to configurin

Let's make it easier as we can.    less env variables will be the best.

**Comment by Djelibeybi** [comment]
Fair enough. Let me work on that.

**Comment by Neilpang** [comment]
Thanks

**Comment by Neilpang** [comment]
Once I learn to use a new tool,   I like it more if it has fewer configurations.

**Comment by Djelibeybi** [comment]
> Once I learn to use a new tool, I like it more if it has fewer configurations.

I agree, but I'm not sure our SDK/CLI team do. :) I'm also working on a version that uses instance principals so if you run `acme.sh` on an OCI instance, it'll auto-configure itself based on data from the OCI metadata service.

**Comment by Neilpang** [comment]
>> I'm also working on a version that uses instance principals so if you run acme.sh on an OCI instance, it'll auto-configure itself based on data from the OCI metadata service.

It would be better, but I would recommend it to be our next goal later.

**Comment by Djelibeybi** [comment]
> It would be better, but I would recommend it to be our next goal later.

Yes, that's my intention too.

**Comment by neilpang** [comment]
Closing this tracking issue for now to keep the issue list clean. It remains the place to report problems with this provider -- if you hit a bug, comment here and the issue will be reopened.
